### PR TITLE
Capturing camera parameters within the URL

### PR DESCRIPTION
**Pull request by maxkuznetsov-actionengine**

### CODE REVIEW

**Review by belom88 — CHANGES_REQUESTED** [review]
We have `src\utils\url-utils.ts` with some logic for `tileset` param. I'd add there functions `viewStateToUrlParams` and `urlParamsToViewState`.

File: src/utils/url-utils.ts [hunk 1]
```diff
@@ -57,3 +57,35 @@ const prepareTilesetUrl = (parsedUrl) => {
     .concat("layers/0");
   return `${parsedUrl.origin}${replacedPathName}${parsedUrl.search}`;
 };
+
+export const viewStateToUrlParams = (viewState, setSearchParams) => {
+  const search = Object.fromEntries(
+    new URLSearchParams(window.location.search)
+  );
+  const { longitude, latitude, pitch, bearing, zoom } = viewState.main;
+  setSearchParams(
```
Comment (belom88) [inline src/utils/url-utils.ts:66]: I don't like having a state setter outside of react. I was about returning some ready to use object and set it in the component.
We can have it here but need to make a TSDoc comments anyway.

File: src/utils/url-utils.ts [hunk 2]
```diff
@@ -57,3 +57,35 @@ const prepareTilesetUrl = (parsedUrl) => {
     .concat("layers/0");
   return `${parsedUrl.origin}${replacedPathName}${parsedUrl.search}`;
 };
+
+export const viewStateToUrlParams = (viewState, setSearchParams) => {
```
Comment (belom88) [inline src/utils/url-utils.ts:61]: Add TSDoc and tests for new functions

File: src/utils/url-utils.ts [hunk 3]
```diff
@@ -57,3 +57,35 @@ const prepareTilesetUrl = (parsedUrl) => {
...
+      latitude,
+      pitch,
+      bearing,
+      zoom,
+    },
+    { replace: true }
+  );
+};
+
+export const urlParamsToViewState = (viewState, setStateUrlViewStateParams) => {
```
Comment (belom88) [inline src/utils/url-utils.ts:79]: Ditto

File: src/utils/url-utils.ts [hunk 4]
```diff
@@ -57,3 +57,35 @@ const prepareTilesetUrl = (parsedUrl) => {
...
+  const urlViewStateParams = {};
+  for (const viewStateParam of search) {
+    if (
+      Object.keys(viewState.main).includes(viewStateParam[0]) &&
+      !isNaN(parseFloat(viewStateParam[1]))
+    ) {
+      urlViewStateParams[viewStateParam[0]] = parseFloat(viewStateParam[1]);
+    }
+  }
+  setStateUrlViewStateParams(urlViewStateParams);
```
Comment (belom88) [inline src/utils/url-utils.ts:90]: Ditto: I prefer returning a value and setting the state in the component.

**Review by belom88 — CHANGES_REQUESTED** [review]
Could you add tests for new functions

File: src/utils/url-utils.ts (hunk 1 above)
Comment (maxkuznetsov-actionengine) [reply src/utils/url-utils.ts:66]: Done

File: src/utils/url-utils.ts (hunk 2 above)
Comment (maxkuznetsov-actionengine) [reply src/utils/url-utils.ts:61]: Done

File: src/utils/url-utils.ts (hunk 3 above)
Comment (maxkuznetsov-actionengine) [reply src/utils/url-utils.ts:79]: Done

File: src/utils/url-utils.ts (hunk 4 above)
Comment (maxkuznetsov-actionengine) [reply src/utils/url-utils.ts:90]: Done

**Review by belom88 — APPROVED** [review]

**Pull request merged**